### PR TITLE
Add Qwen2/2.5-VL unified implementation

### LINKED ISSUE #167: Extend `VisionAddOn` Pattern to Qwen2.5VL
Currently, the only multi-modal models that have been migrated to the "unified" architecture are Gemma3 and Pixtral:

https://github.com/lmstudio-ai/mlx-engine/blob/ecc2cf48e449478784ffd21f1d16745e11647e2f/mlx_engine/model_kit/model_kit.py#L35-L38

Extending this pattern to Qwen2.5VL/Qwen2VL is desired.

Relevant `mlx-vlm` components:
- https://github.com/Blaizzy/mlx-vlm/tree/2068970094c78878c77fd78677d1316933562ade/mlx_vlm/models/qwen2_5_vl
- https://github.com/Blaizzy/mlx-vlm/tree/main/mlx_vlm/models/qwen2_vl

Relevant `mlx-lm` components:
- https://github.com/ml-explore/mlx-lm/blob/77edf17bc0bf7c9313e0b970490db86a4f64bee4/mlx_lm/models/qwen2.py

This will likely look like:
1. Ensure Qwen2.5VL text model architecture is implemented correctly in `mlx-lm` (including MRoPE, see https://arxiv.org/abs/2502.13923 for details and [Apply PR #319 fixes to Qwen 2.5VL position id #349](https://github.com/Blaizzy/mlx-vlm/pull/349) for mlx-vlm in progress work)
2. Implement `Qwen2_5_VLVisionAddOn` and wire it in `ModelKit`
3. Ensure Qwen2.5VL tests in `mlx-engine` still pass

### PR DESCRIPTION
**Pull request by vincentamato**
Resolves #167 

Requires a sibling PR to `mlx-lm` to bring in the `qwen2_vl` language model [ml-explore/mlx-lm#384](https://github.com/ml-explore/mlx-lm/pull/384)

This implementation is cleaner and more consistent with the current codebase than my earlier PR #210 . No changes are made to the core logic of `ModelKit`. All updates are contained within the new vision add-on:

-  During initialization, the `load_vision_addon` function found in the other vision add-ons is supplemented due to Qwen2/2.5-VL not having a multi-modal projector.
- When computing embeddings, the `common_process_prompt_with_images` function found in the other vision add-ons is  supplemented due to the way Qwen2/2.5-VL models handle padding with multi-image prompts.

Tests continue to pass with this addition.

### CODE REVIEW

File: mlx_engine/model_kit/vision_add_ons/qwen2_vl.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+    """Container for Qwen2-VL vision components."""
+
+    def __init__(self, vision_tower: nn.Module):
+        super().__init__()
+        self.vision_tower = vision_tower
+
+
+class Qwen2_VLVisionAddOn(BaseVisionAddOn):
+    """
+    Vision add-on for Qwen2-VL and Qwen2.5-VL models.
```
Comment (mattjcly) [inline mlx_engine/model_kit/vision_add_ons/qwen2_vl.py:32]: Curious, what went into the design decision to combine these two into a single `VisionAddOn` rather than make two separate ones?

File: mlx_engine/model_kit/vision_add_ons/qwen2_vl.py [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+    """
+
+    def __init__(self, model_path: Path):
+        """Initialize Qwen2_VLVisionAddOn with vision components loaded from the given path."""
+        super().__init__()
+
+        # Determine model type from config to select appropriate classes
```
Comment (mattjcly) [inline mlx_engine/model_kit/vision_add_ons/qwen2_vl.py:39]: Nit: could consider use of a helper function called something like `load_qwen_vision_addon` that encapsulates the below and makes clear it's parallel to `load_vision_addon` from `load_utils.py`

File: mlx_engine/model_kit/vision_add_ons/qwen2_vl.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+            self.config.video_token_id,
+            hidden_states,
+            input_embeddings,
+            input_ids,
+        )
+
+        # Remove batch dimension
+        return input_ids.squeeze(0), final_inputs_embeds.squeeze(0)
+
+    def _resize_images(self, images, max_size=(1000, 1000)):
```
Comment (mattjcly) [inline mlx_engine/model_kit/vision_add_ons/qwen2_vl.py:167]: Did you see the existence of this function? 
https://github.com/lmstudio-ai/mlx-engine/blob/7a16c21ade4503e6614d583b8c72c3e66985d56e/mlx_engine/utils/image_utils.py#L18

Any thoughts on using it or adopting it in this file to reduce duplicate code?

**Review by mattjcly — COMMENTED** [review]
@vincentamato Very cool, thank you for the contribution! Will be able to look more in-depth tomorrow, here are just an initial batch of comments/questions about the implementation

File: mlx_engine/model_kit/vision_add_ons/qwen2_vl.py (hunk 2 above)
Comment (vincentamato) [reply mlx_engine/model_kit/vision_add_ons/qwen2_vl.py:39]: Great idea, will do

File: mlx_engine/model_kit/vision_add_ons/qwen2_vl.py (hunk 3 above)
Comment (vincentamato) [reply mlx_engine/model_kit/vision_add_ons/qwen2_vl.py:167]: The problem is that adding padding to the images breaks multi-image support for these models. Maybe I modify that function to variably pad images?

**Comment by vincentamato** [comment]
@mattjcly My latest commit addressed some of the issues you brought up

File: mlx_engine/model_kit/vision_add_ons/qwen2_vl.py (hunk 1 above)
Comment (mattjcly) [reply mlx_engine/model_kit/vision_add_ons/qwen2_vl.py:18]: Totally seems reasonable if "downstream of that everything is the same" is verified to be the correct behavior. For example, I know there are minor differences in how MRoPE between the two are implemented.

I added a comment to mlx-lm pr to the tune of this: https://github.com/ml-explore/mlx-lm/pull/384#discussion_r2288410031

Very well may be the case that downstream everything is indeed the same, but want to make sure we've done the proper research to make the sure of it (even if it means in practice we just understand and acknowledge gaps like "video input won't work correctly but thats fine for now").

Does that make sense?

**Comment by mattjcly** [comment]
@vincentamato Thanks for the changes. I would say we're just blocked on the mlx-lm merge here (which I see some new changes went in recently and just waiting on review), after which will do a final pass here.

**Comment by vincentamato** [comment]
Hey @neilmehta24, good catch. The slowdown was being caused by a dtype mismatch between `pixel_values` and `input_embeddings`. Without explicit conversion, MLX was performing implicit type conversions throughout the vision tower and embedding merge operations, resulting in the slowdown. I've added a dtype alignment check that brings performance parity with the main branches.

**Review by neilmehta24 — COMMENTED** [review]
Thanks for the generation speed fix. I have some minor comments about some of the import statements. After addressing, this should be good to merge.

**Review by neilmehta24 — APPROVED** [review]
LGTM, thank you for the contribution!
